packages/core/src/routes/phrase.ts

```diff
@@ -0,0 +1,130 @@
+import { LanguageKey, languageKeyGuard } from '@logto/core-kit';
+import resource from '@logto/phrases-ui';
+import en from '@logto/phrases-ui/lib/locales/en';
+import { CustomPhrases, Translation } from '@logto/schemas';
+import { adminConsoleApplicationId, adminConsoleSignInExperience } from '@logto/schemas/lib/seeds';
+import { notFalsy } from '@silverhand/essentials';
+import deepmerge from 'deepmerge';
+import { ResourceLanguage } from 'i18next';
+import { errors, Provider } from 'oidc-provider';
+
+import detectLanguage from '@/i18n/detect-language';
+import koaGuard from '@/middleware/koa-guard';
+import { findAllCustomLanguageKeys, findCustomPhraseByLanguageKey } from '@/queries/custom-phrase';
+import { findDefaultSignInExperience } from '@/queries/sign-in-experience';
+import assertThat from '@/utils/assert-that';
+
+import { AnonymousRouter } from './types';
+
+const getLanguageInfo = async (applicationId: unknown) => {
+  if (applicationId === adminConsoleApplicationId) {
+    return adminConsoleSignInExperience.languageInfo;
+  }
+
+  const { languageInfo } = await findDefaultSignInExperience();
+
+  return languageInfo;
+};
+
+const isBuiltInLanguageKey = (key: string): key is LanguageKey =>
+  Object.keys(resource).includes(key) && languageKeyGuard.safeParse(key).success;
+
+const isFullyTranslated = (fullTranslation: Translation, targetTranslation: Translation) => {
+  const keys = Object.keys(fullTranslation);
+  const targetKeys = new Set(Object.keys(targetTranslation));
+
+  if (keys.length !== targetKeys.size) {
+    return false;
+  }
+
+  if (!keys.every((key) => targetKeys.has(key) && notFalsy(targetTranslation[key]))) {
+    return false;
+  }
+
+  for (const key of keys) {
+    if (
+      typeof fullTranslation[key] === 'object' &&
+      // eslint-disable-next-line no-restricted-syntax
+      !isFullyTranslated(fullTranslation[key] as Translation, targetTranslation[key] as Translation)
+    ) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+const finalDefaultLanguage = 'en';
+// Ensure the final default language phrase contains a built-in phrase (that means it must be fully translated).
+assertThat(
+  Object.keys(resource).includes(finalDefaultLanguage),
+  new Error(`The final default language "${finalDefaultLanguage}" must contain a built-in phrase.`)
+);
+
+const getPhrase = async (
```

Comment:
nit: we can extract this function, and test it directly, it would be more clear compare to the long route testing code